question-servers/freeform.js

```diff
@@ -1255,4 +1238,62 @@ module.exports = {
             callback(null, context);
         });
     },
+
+    getCacheKey: function(course, data, context, callback) {
+        courseUtil.getOrUpdateCourseCommitHash(course, (err, commitHash) => {
+            if (ERR(err, callback)) return;
+            const dataHash = hash('sha1').update(JSON.stringify({data, context})).digest('base64');
+            callback(null, `${commitHash}-${dataHash}`);
+        });
+    },
+
+    getCachedDataOrCompute: function(course, data, context, computeFcn, processFcn, errorFcn) {
```

Comment:
Nice refactor! Almost makes me wish there was more we could cache with the freeform server

Reply:
Yes! Caching is awesome! In typical usage, more than half of all question HTML is served from the cache. @nwalters512 wrote this cache code and convinced me that it was actually worth running 😄